File: src/BaseHmiTypes/Scripts/HmiScriptCommand.cs
namespace BaseHmiTypes.Scripts;

public abstract class HmiScriptCommand
{
    private readonly List<HmiScriptCommandArgument> _arguments = new List<HmiScriptCommandArgument>();

    public abstract string CommandName { get; }

    public virtual string CommandKey => CommandName;

    public IList<HmiScriptCommandArgument> Arguments => _arguments;

    protected HmiScriptCommandArgument AddArgument(string propertyName, string parameterName, HmiScriptCommandParameterDirection direction)
    {
        var argument = new HmiScriptCommandArgument
        {
            PropertyName = propertyName,
            ParameterName = parameterName,
            Direction = direction
        };

        _arguments.Add(argument);
        return argument;
    }
}

public sealed class HmiScriptCommandArgument
{
    public string PropertyName { get; set; } = string.Empty;

    public string ParameterName { get; set; } = string.Empty;

    public HmiScriptCommandParameterDirection Direction { get; set; } = HmiScriptCommandParameterDirection.In;

    public HmiScriptArgumentValue? Value { get; set; }
}

public sealed class HmiScriptArgumentValue
{
    public HmiScriptArgumentSourceKind SourceKind { get; set; } = HmiScriptArgumentSourceKind.Unknown;

    public string? RawValue { get; set; }

    public object? Value { get; set; }

    public string? ObjectType { get; set; }

    public string? ValueType { get; set; }
}

public enum HmiScriptArgumentSourceKind
{
    Unknown,
    Constant,
    ObjectReference,
    Integer,
    Double,
    Enum,
    Text
}

File: typescript/base-hmi-types/src/scripts/HmiScriptCommand.ts
export abstract class HmiScriptCommand {
  private readonly argumentsCore: HmiScriptCommandArgument[] = [];

  abstract get commandName(): string;

  get commandKey(): string {
    return this.commandName;
  }

  get arguments(): HmiScriptCommandArgument[] {
    return this.argumentsCore;
  }

  protected addArgument(propertyName: string, parameterName: string, direction: HmiScriptCommandParameterDirection): HmiScriptCommandArgument {
    const argument = new HmiScriptCommandArgument();
    argument.propertyName = propertyName;
    argument.parameterName = parameterName;
    argument.direction = direction;
    this.argumentsCore.push(argument);
    return argument;
  }
}

export class HmiScriptCommandArgument {
  propertyName = "";
  parameterName = "";
  direction: HmiScriptCommandParameterDirection = HmiScriptCommandParameterDirection.In;
  value?: HmiScriptArgumentValue;
}

export class HmiScriptArgumentValue {
  sourceKind: HmiScriptArgumentSourceKind = HmiScriptArgumentSourceKind.Unknown;
  rawValue?: string;
  value?: unknown;
  objectType?: string;
  valueType?: string;
}

export enum HmiScriptArgumentSourceKind {
  Unknown = "Unknown",
  Constant = "Constant",
  ObjectReference = "ObjectReference",
  Integer = "Integer",
  Double = "Double",
  Enum = "Enum",
  Text = "Text",
}

export enum HmiScriptCommandParameterDirection {
  In = "In",
  Out = "Out",
  Return = "Return",
}
